### ISSUE TITLE
Unbounded Memory leak when using tf.py_function in tf.data.Dataset.map()

### ISSUE THREAD

**Issue by Pyrestone** [tensorflow/tensorflow#61344]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

v2.13.0-rc2-7-g1cb1a030a62 2.13.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 20.04, Google Colab

### Python version

3.8

### CUDA/cuDNN version

11.8 / 8.6

### GPU model and memory

various, e.g. 2080ti, 3080ti mobile, Colab T4

### Current behavior?

Using tf.py_function in a function that is applied to a tf.data.Dataset via its map() function causes a (C++-level) memory leak.

In my real training with more complex code inside the py_function, this lead to the python script eventually consuming upwards of 30 GB of RAM during a model.fit() loop, despite taking less that 3GB of RAM during the initial epoch.

tf.py_function also more generally causes memory leaks in all kinds of places. See the flags at the top of the linked Collab for details.

### Standalone code to reproduce the issue

```shell
see Collab: https://colab.research.google.com/drive/1auVJPyHApl4__4FF-rV3xNcJrqYZc38R?usp=sharing

Iterating through a dataset with a tf.py_function in it causes unbounded linear memory consumption growth.
```

### Relevant log output

```shell
**Batch 0**
Memory usage: 1732120576
Delta: 1651.88 MiB
**Batch 200**
Memory usage: 1736859648
Delta: 4.52 MiB
**Batch 400**
Memory usage: 1740644352
Delta: 3.61 MiB
**Batch 600**
Memory usage: 1744699392
Delta: 3.87 MiB
Average Delta since start: 3.87 MiB/iteration
Estimated growth per 1000 steps: 19.34 MiB
**Batch 800**
Memory usage: 1748750336
Delta: 3.86 MiB
Average Delta since start: 3.87 MiB/iteration
Estimated growth per 1000 steps: 19.33 MiB
**Batch 1000**
Memory usage: 1752805376
Delta: 3.87 MiB
Average Delta since start: 3.87 MiB/iteration
Estimated growth per 1000 steps: 19.33 MiB
**Batch 1200**
Memory usage: 1757401088
Delta: 4.38 MiB
Average Delta since start: 4.00 MiB/iteration
Estimated growth per 1000 steps: 19.98 MiB
**Batch 1400**
Memory usage: 1761456128
Delta: 3.87 MiB
Average Delta since start: 3.97 MiB/iteration
Estimated growth per 1000 steps: 19.85 MiB
**Batch 1600**
Memory usage: 1765240832
Delta: 3.61 MiB
Average Delta since start: 3.91 MiB/iteration
Estimated growth per 1000 steps: 19.55 MiB
**Batch 1800**
Memory usage: 1769025536
Delta: 3.61 MiB
Average Delta since start: 3.87 MiB/iteration
Estimated growth per 1000 steps: 19.33 MiB
**Batch 2000**
Memory usage: 1773621248
Delta: 4.38 MiB
Average Delta since start: 3.93 MiB/iteration
Estimated growth per 1000 steps: 19.66 MiB
**Batch 2200**
Memory usage: 1777676288
Delta: 3.87 MiB
Average Delta since start: 3.92 MiB/iteration
Estimated growth per 1000 steps: 19.62 MiB
**Batch 2400**
Memory usage: 1781731328
Delta: 3.87 MiB
Average Delta since start: 3.92 MiB/iteration
Estimated growth per 1000 steps: 19.59 MiB
**Batch 2600**
Memory usage: 1785786368
Delta: 3.87 MiB
Average Delta since start: 3.91 MiB/iteration
Estimated growth per 1000 steps: 19.57 MiB
```

**Comment by Pyrestone** [comment]
see also: #35084 and #51839 (both closed, but the issue re-appeared in 2.13.0)

**Comment by SuryanarayanaY** [comment]
Hi @Pyrestone ,

I have replicated the reported behaviour and attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/84a602e1d03dd977317080607592bd0a/61344_-tf_memory_leak_py_function.ipynb) for reference. 

I can observe memory increased from  batch:0 to batch:3600. Needs to be investigated.

**Comment by Pyrestone** [comment]
Note for people finding this issue in the meantime:

As a temporary (or permanent I suppose) **workaround** is using `tf.numpy_function` instead of `tf.py_function` in tf.data.Dataset.map().  

It behaves very similarly, except the function then receives and returns numpy arrays instead of eager tensors.  
The only other notable difference is that `tf.numpy_function` doesn't support gradient computation (which probably shouldn't matter in a dataset.map() call).

**Comment by hyliker** [comment]
@Pyrestone Thanks for your solution.

**Comment by tilakrayal** [comment]
@Pyrestone,
COuld you please try to check in the latest tensorflow v2.17 where it contains the Keras3.0 version and also tried using tf.numpy_function instead of tf.py_function & observed the code was executed as expected. Thank you!

**Comment by Pyrestone** [comment]
I ran some tests with 2.13,2.15, and 2.17, apparently the memory leak issue was actually fixed somewhere between 2.13 and 2.15.0  
However, between 2.15.0 and 2.17, something in `tf.data` or `tf.py_function` became noticeably slower, (less than 1800 steps per hour instead of 3000 before).

*2.13.0*:
```txt
**Batch 0**
Time since start: 0.00 minutes
Memory usage: 780132352
Delta: 743.99 MiB
**Batch 200**
Time since start: 4.13 minutes
Memory usage: 833122304
Delta: 50.54 MiB
**Batch 400**
Time since start: 8.54 minutes
Memory usage: 835723264
Delta: 2.48 MiB
**Batch 600**
Time since start: 12.62 minutes
Memory usage: 838123520
Delta: 2.29 MiB
Average Delta since start: 2.29 MiB/iteration
Estimated growth per 1000 steps: 11.45 MiB
**Batch 800**
Time since start: 16.95 minutes
Memory usage: 840826880
Delta: 2.58 MiB
Average Delta since start: 2.43 MiB/iteration
Estimated growth per 1000 steps: 12.17 MiB
**Batch 1000**
Time since start: 21.28 minutes
Memory usage: 843534336
Delta: 2.58 MiB
Average Delta since start: 2.48 MiB/iteration
Estimated growth per 1000 steps: 12.42 MiB
**Batch 1200**
Time since start: 25.44 minutes
Memory usage: 879480832
Delta: 34.28 MiB
Average Delta since start: 10.43 MiB/iteration
Estimated growth per 1000 steps: 52.16 MiB
**Batch 1400**
Time since start: 29.31 minutes
Memory usage: 882413568
Delta: 2.80 MiB
Average Delta since start: 8.91 MiB/iteration
Estimated growth per 1000 steps: 44.53 MiB
**Batch 1600**
Time since start: 33.10 minutes
Memory usage: 885178368
Delta: 2.64 MiB
Average Delta since start: 7.86 MiB/iteration
Estimated growth per 1000 steps: 39.30 MiB
**Batch 1800**
Time since start: 36.89 minutes
Memory usage: 887635968
Delta: 2.34 MiB
Average Delta since start: 7.07 MiB/iteration
Estimated growth per 1000 steps: 35.36 MiB
**Batch 2000**
Time since start: 40.46 minutes
Memory usage: 890339328
Delta: 2.58 MiB
Average Delta since start: 6.51 MiB/iteration
Estimated growth per 1000 steps: 32.55 MiB
**Batch 2200**
Time since start: 44.94 minutes
Memory usage: 875814912
Delta: -13.85 MiB
Average Delta since start: 4.25 MiB/iteration
Estimated growth per 1000 steps: 21.24 MiB
**Batch 2400**
Time since start: 49.37 minutes
Memory usage: 861831168
Delta: -13.34 MiB
Average Delta since start: 2.49 MiB/iteration
Estimated growth per 1000 steps: 12.45 MiB
**Batch 2600**
Time since start: 53.75 minutes
Memory usage: 864575488
Delta: 2.62 MiB
Average Delta since start: 2.50 MiB/iteration
Estimated growth per 1000 steps: 12.51 MiB
**Batch 2800**
Time since start: 58.20 minutes
Memory usage: 883744768
Delta: 18.28 MiB
Average Delta since start: 3.82 MiB/iteration
Estimated growth per 1000 steps: 19.08 MiB
```

*2.15.0*:
```txt
**Batch 0**
Time since start: 0.00 minutes
Memory usage: 881840128
Delta: 840.99 MiB
**Batch 200**
Time since start: 4.06 minutes
Memory usage: 898949120
Delta: 16.32 MiB
**Batch 400**
Time since start: 8.07 minutes
Memory usage: 898904064
Delta: -0.04 MiB
**Batch 600**
Time since start: 12.21 minutes
Memory usage: 898744320
Delta: -0.15 MiB
Average Delta since start: -0.15 MiB/iteration
Estimated growth per 1000 steps: -0.76 MiB
**Batch 800**
Time since start: 16.54 minutes
Memory usage: 898949120
Delta: 0.20 MiB
Average Delta since start: 0.02 MiB/iteration
Estimated growth per 1000 steps: 0.11 MiB
**Batch 1000**
Time since start: 20.82 minutes
Memory usage: 898805760
Delta: -0.14 MiB
Average Delta since start: -0.03 MiB/iteration
Estimated growth per 1000 steps: -0.16 MiB
**Batch 1200**
Time since start: 25.35 minutes
Memory usage: 898928640
Delta: 0.12 MiB
Average Delta since start: 0.01 MiB/iteration
Estimated growth per 1000 steps: 0.03 MiB
**Batch 1400**
Time since start: 29.74 minutes
Memory usage: 899055616
Delta: 0.12 MiB
Average Delta since start: 0.03 MiB/iteration
Estimated growth per 1000 steps: 0.14 MiB
**Batch 1600**
Time since start: 34.09 minutes
Memory usage: 899223552
Delta: 0.16 MiB
Average Delta since start: 0.05 MiB/iteration
Estimated growth per 1000 steps: 0.25 MiB
**Batch 1800**
Time since start: 38.96 minutes
Memory usage: 899223552
Delta: 0.00 MiB
Average Delta since start: 0.04 MiB/iteration
Estimated growth per 1000 steps: 0.22 MiB
**Batch 2000**
Time since start: 42.37 minutes
Memory usage: 899203072
Delta: -0.02 MiB
Average Delta since start: 0.04 MiB/iteration
Estimated growth per 1000 steps: 0.18 MiB
**Batch 2200**
Time since start: 45.74 minutes
Memory usage: 899182592
Delta: -0.02 MiB
Average Delta since start: 0.03 MiB/iteration
Estimated growth per 1000 steps: 0.15 MiB
**Batch 2400**
Time since start: 49.86 minutes
Memory usage: 932950016
Delta: 32.20 MiB
Average Delta since start: 3.25 MiB/iteration
Estimated growth per 1000 steps: 16.23 MiB
**Batch 2600**
Time since start: 54.24 minutes
Memory usage: 899493888
Delta: -31.91 MiB
Average Delta since start: 0.05 MiB/iteration
Estimated growth per 1000 steps: 0.26 MiB
**Batch 2800**
Time since start: 58.58 minutes
Memory usage: 899534848
Delta: 0.04 MiB
Average Delta since start: 0.05 MiB/iteration
Estimated growth per 1000 steps: 0.25 MiB
```

*2.17.0*:
```txt
**Batch 0**
Time since start: 0.00 minutes
Memory usage: 883638272
Delta: 842.70 MiB
**Batch 200**
Time since start: 7.98 minutes
Memory usage: 934363136
Delta: 48.38 MiB
**Batch 400**
Time since start: 15.04 minutes
Memory usage: 900763648
Delta: -32.04 MiB
**Batch 600**
Time since start: 21.68 minutes
Memory usage: 900784128
Delta: 0.02 MiB
Average Delta since start: 0.02 MiB/iteration
Estimated growth per 1000 steps: 0.10 MiB
**Batch 800**
Time since start: 28.35 minutes
Memory usage: 901111808
Delta: 0.31 MiB
Average Delta since start: 0.17 MiB/iteration
Estimated growth per 1000 steps: 0.83 MiB
**Batch 1000**
Time since start: 35.16 minutes
Memory usage: 901328896
Delta: 0.21 MiB
Average Delta since start: 0.18 MiB/iteration
Estimated growth per 1000 steps: 0.90 MiB
**Batch 1200**
Time since start: 41.80 minutes
Memory usage: 934793216
Delta: 31.91 MiB
Average Delta since start: 8.11 MiB/iteration
Estimated growth per 1000 steps: 40.57 MiB
**Batch 1400**
Time since start: 49.42 minutes
Memory usage: 901423104
Delta: -31.82 MiB
Average Delta since start: 0.13 MiB/iteration
Estimated growth per 1000 steps: 0.63 MiB
**Batch 1600**
Time since start: 57.31 minutes
Memory usage: 901312512
Delta: -0.11 MiB
Average Delta since start: 0.09 MiB/iteration
Estimated growth per 1000 steps: 0.44 MiB
**Batch 1800**
Time since start: 65.39 minutes
Memory usage: 901586944
Delta: 0.26 MiB
Average Delta since start: 0.11 MiB/iteration
Estimated growth per 1000 steps: 0.56 MiB
**Batch 2000**
Time since start: 73.38 minutes
Memory usage: 901701632
Delta: 0.11 MiB
Average Delta since start: 0.11 MiB/iteration
Estimated growth per 1000 steps: 0.56 MiB
```

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
